sympy/matrices/common.py

```diff
@@ -2535,6 +2536,8 @@ def __abs__(self):
 
     @call_highest_priority('__radd__')
     def __add__(self, other):
+        if isinstance(other, NDimArray):
+            return NotImplemented
         """Return self + other, raising ShapeError if shapes don't match."""
```

Comment:
There should not be code before the docstring.

Also this should be phrased as:
```
if not isinstance(other, ...):
     return NotImplemented
```
The method should return `NotImplemented` unless it explicitly knows how to interoperate with other object.

Reply:
In this case I think it should be if isinstance(other, NdimArray) because matrix class is meant to work work with all objects having similar shape, if not isinstance will make matrix class work only with matrix not any other object of similar shape. I think only Matrix and Array addition should return NotImplemented

Reply:
What does "other object of similar shape" mean?

In general `NotImplemented` should be returned except for a specific list of allowed types.

Reply:
Can you please list the allowed types?

Reply:
> What does "other object of similar shape" mean?

Matrix like object